### PR TITLE
EOS-16875: Modify s3_setup shell script to Python - Test phase

### PR DESCRIPTION
**Pull request by pratyush-seagate**
commit c229a7be6f9111326d49d60883be102e9ba30454
Author: pratyush-seagate <pratyush.k.khan@seagate.com>
Date:   Fri Feb 5 07:04:23 2021 -0700

        EOS-16875: Modify s3_setup shell script to Python - Test phase
        Changes in files:
                modified:   installhelper.sh
                modified:   scripts/provisioning/s3setup/s3setup/cortx_s3_setup.py
                renamed:    s3-sanity-test.sh -> scripts/s3-sanity/s3-sanity-test.sh
        Description:
                1. Update existing s3-sanity-test.sh, to make it work with current mini-provisioner environment
                2. Add -d, -f and -p input params in s3-sanity-test.sh
                3. Remove redundant and obsolete version of the s3-sanity-test.sh
                4. Add new option "test" in s3setup
                5. cleanup option param fix in s3setup
        Test:
                Unit test in mini-provisioner VM setup. Following tests passed:
                s3setup test --ldappasswd  "abcd"
                s3setup test --ldappasswd  "abcd" --data_sanity --s3cmdcfg /tmp/s3cmdcfg

                scripts/s3-sanity/s3-sanity-test.sh -p ldappasswd
                scripts/s3-sanity/s3-sanity-test.sh -p ldappasswd -d -f /tmp/s3cmdcfg

                Note: To run the data tests using s3cmd, --s3cmdcfg must be created before hand with valid values and mini-provisioner single VM setup should complete steps needed for "Start s3server and motr for I/O"
                Manual mini-provisioner single VM deployment and individual scripts/s3-sanity/s3-sanity-test.sh run.

    Signed-off-by: pratyush-seagate <pratyush.k.khan@seagate.com>

### CODE REVIEW

**Review by shahidSTX — APPROVED** [review]

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 1]
```diff
@@ -17,81 +18,221 @@
...
+else
+    die_with_error "\ns3iamcli not installed\nPlease install s3iamcli using rpm built from <s3server repo>/rpms/s3iamcli/ or from rpm cortx-s3iamcli"
+fi
+
+if [ -z "$ldappasswd" ]
+then
+  die_with_error "Please provide -p <ldappasswd>"
+fi
+
+systemctl restart s3authserver
```
Comment (amitwac2608) [inline scripts/s3-sanity/s3-sanity-test.sh:172]: why this is needed ?

Comment (amitwac2608) [reply scripts/s3-sanity/s3-sanity-test.sh:172]: Can we just check if s3authserver is up and running ?

**Comment by gowthamchinna** [comment]
run mini-provisioning

File: scripts/s3-sanity/s3-sanity-test.sh [hunk 2]
```diff
@@ -17,81 +18,221 @@
...
+data_path_test=false
+s3cmdconfigfile="/root/.s3cfg"
+
+set -e
+set -p
+
+USAGE="USAGE: bash $(basename "$0")
 Run S3 sanity test.
+
+PRE-REQUISITES: s3cmd and s3iamcli need to be present
```
[the remark this answers is not in the document]
Comment (pratyush-seagate) [reply scripts/s3-sanity/s3-sanity-test.sh:38]: Yes, but for s3cmd, user needs to pass the config file manually using "-f" flag. Due to restrictions, the config file was not added in the rpm and can not be part of the deployment automatically. I was suggested by Basav to explore of s3cmd can be run without config file. I think we can explore that in a later change as this change primarily targets the use of s3iamcli in the test phase.

File: scripts/provisioning/s3setup/s3setup/cortx_s3_setup.py [hunk 3]
```diff
@@ -160,3 +182,17 @@ def run(self):
 
       rc = self.validate_pre_requisites(rpms=preqs_conf_json['rpms'], services=preqs_conf_json['services'], pip3s=preqs_conf_json['pip3s'])
       exit(not rc)
+
+    if "test" in args.action:
+      if args.ldappasswd == None:
+        print("Invalid input, provide --ldappasswd for sanity test")
+        exit (-2)
+      if args.data_sanity:
+        if args.s3cmdcfg == None:
```
[the remark this answers is not in the document]
Comment (pratyush-seagate) [reply scripts/provisioning/s3setup/s3setup/cortx_s3_setup.py:191]: Both the actual shell script and python callers are actually consistent. If s3cmd is used, i.e. -d or --data_sanity flags are passed, then the s3cmd config file as mandatory option. 
If the s3cmd is not used, then s3cmd config file as not mandatory in both places.

File: scripts/s3-sanity/s3-sanity-test.sh (hunk 1 above)
Comment (pratyush-seagate) [reply scripts/s3-sanity/s3-sanity-test.sh:172]: @AjinkyaDhumal  suggested that the check is not needed for the time being.

File: scripts/provisioning/s3setup/s3setup/cortx_s3_setup.py [hunk 4]
```diff
@@ -106,40 +107,61 @@ def validate_pre_requisites(self, rpms: list = None, pip3s: list = None, service
...
 
+  def sanity_test(self, ldappasswd: str, dataflag: bool = False, s3cmdcfg: str = None):
+    try:
+      cmd = ['/opt/seagate/cortx/s3/scripts/s3-sanity-test.sh', f'-p {ldappasswd}']
+      if dataflag:
+        if not s3cmdcfg:
+          print("Please provide s3cmdcfg")
+          return False
+        cmd.append('-d')
+        cmd.append(f'-f {s3cmdcfg}')
```
[the remark this answers is not in the document]
Comment (pratyush-seagate) [reply scripts/provisioning/s3setup/s3setup/cortx_s3_setup.py:118]: s3-sanity-test.sh already has this validation check. Anyways I'll also add it in the python version.

**Comment by pratyush-seagate** [comment]
From http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/cortx-s3server/203/console:

10:50:18  TASK [s3server : [mini_provisioning] : S3 Sanity Test] *************************
10:50:31  fatal: [srvnode-1]: FAILED! => changed=true 
10:50:31    cmd: /opt/seagate/cortx/s3/bin/s3_setup test --config "json:///****/confstore.json"
10:50:31    delta: '0:00:12.553375'
10:50:31    end: '2021-02-08 22:20:31.768865'
10:50:31    failed_when_result: true
10:50:31    msg: non-zero return code
10:50:31    rc: 1
10:50:31    start: '2021-02-08 22:20:19.215490'
10:50:31    stderr: ''
10:50:31    stderr_lines: <omitted>
10:50:31    stdout: |-
10:50:31      INFO: Get openldap passwords from confstore..
10:50:31      ERROR:\n ['/opt/seagate/cortx/s3/scripts/s3-sanity-test.sh', '-p ****'] failed with err: b'', out: b'\nCheck s3iamcli...OK\nFailed to delete user.\nUnable to locate credentials\n\nAccount cannot be deleted.\nAn error occurred (InvalidAccessKeyId) : The AWS access key Id you provided does not exist in our records.\n\n/****/.sgs3iamcli/config.yaml\n\n\n*** S3 Sanity ***\n\n\n**** Create Account *******\n\nFailed to delete user.\nUnable to locate credentials\n\nAccount cannot be deleted.\nAn error occurred (InvalidAccessKeyId) : The AWS access key Id you provided does not exist in our records.\n\n/****/.sgs3iamcli/config.yaml\n', ret: 1 \n
10:50:31      ERROR: s3setup test failed, exit.
10:50:31    stdout_lines: <omitted>
10:50:31  
10:50:31  PLAY RECAP *********************************************************************
10:50:31  localhost                  : ok=4    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
10:50:31  srvnode-1                  : ok=52   changed=46   unreachable=0    failed=1    skipped=2    rescued=0    ignored=0   
10:50:31  
10:50:31  FATAL: command execution failed

I tried to repro this on local VM and was able to reproduce it. It seems a regression because these steps earlier used to work fine.  Note that final "s3iamcli deleteaccount" is failing in manual steps.

s3iamcli createaccount -n SanityAccountToDeleteAfterUse -e SanityAccountToDeleteAfterUse@sanitybucket.com --ldapuser sgiamadmin --ldappasswd ldapadmin --no-ssl
AccountId = 679414704002, CanonicalId = 266d16c8212847f6a065bbb7e05a3a7d6e42ad6c5f114bcc8fb6f85c036595fe, RootUserName = root, AccessKeyId = AKIAWD_yIzU3Sr2y4NUgJ7EHZw, SecretKey = MGDI/qTAF8Dnq2Wr1EVPGCnp8cFcLR313fY6Di4t

s3iamcli CreateUser -n SanityUserToDeleteAfterUse --access_key AKIAWD_yIzU3Sr2y4NUgJ7EHZw --secret_key MGDI/qTAF8Dnq2Wr1EVPGCnp8cFcLR313fY6Di4t --no-ssl
UserId = AIDA535AB5C1318142719, ARN = arn:aws:iam::679414704002:user/SanityUserToDeleteAfterUse, Path = /


s3iamcli deleteuser -n SanityUserToDeleteAfterUse --access_key AKIAWD_yIzU3Sr2y4NUgJ7EHZw --secret_key MGDI/qTAF8Dnq2Wr1EVPGCnp8cFcLR313fY6Di4t --no-ssl
User deleted.


s3iamcli deleteaccount -n SanityAccountToDeleteAfterUse --access_key AKIAWD_yIzU3Sr2y4NUgJ7EHZw --secret_key MGDI/qTAF8Dnq2Wr1EVPGCnp8cFcLR313fY6Di4t --no-ssl
Account cannot be deleted.
An error occurred (InternalFailure) : The request processing has failed because of an unknown error, exception or failure.

**Comment by pratyush-seagate** [comment]
The mini-rovisioning reports sanity test fails, however s3's sanity is actually passing.

22:26:20  TASK [s3server : [mini_provisioning] : S3 Sanity Test] *************************
22:26:34  fatal: [srvnode-1]: FAILED! => changed=true 
22:26:34    cmd: /opt/seagate/cortx/s3/bin/s3_setup test --config "json:///****/confstore.json"
22:26:34    delta: '0:00:13.536060'
22:26:34    end: '2021-02-09 09:56:34.521710'
22:26:34    failed_when_result: true
22:26:34    rc: 0
22:26:34    start: '2021-02-09 09:56:20.985650'
22:26:34    stderr: ''
22:26:34    stderr_lines: <omitted>
22:26:34    stdout: |-
**22:26:34      INFO: Get openldap passwords from confstore..
22:26:34      PASS: s3setup test passed.**
22:26:34    stdout_lines: <omitted>
22:26:34  
22:26:34  PLAY RECAP *********************************************************************
22:26:34  localhost                  : ok=4    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
22:26:34  srvnode-1                  : ok=52   changed=46   unreachable=0    failed=1    skipped=2    rescued=0    ignored=0   
22:26:34  

Not sure what is actually happening.

**Comment by gowthamchinna** [comment]
@pratyush-seagate Seems test result text changed.

we are checking the text "**PASS: S3-Sanity test passed**" in the s3_setup test but found  "**PASS: s3setup test passed**". Please maintain consistency in the test script return data.

**Comment by pratyush-seagate** [comment]
> 
> 
> @pratyush-seagate Seems test result text changed.
> 
> we are checking the text "**PASS: S3-Sanity test passed**" in the s3_setup test but found "**PASS: s3setup test passed**". Please maintain consistency in the test script return data.

Thanks @gowthamchinna for this crucial information. I did not know this, and was trying to figure out why it was failing.

**Pull request closed without merging**